Adds escalation step to notify a team

xssfox opened the pull request:
# What this PR does
Adds a "Notify Team" escalation step which will notify all users in the assigned grafana team.

<img width="966" alt="Screenshot 2023-12-01 at 10 35 47 am" src="https://github.com/grafana/oncall/assets/59047654/7c5181fc-5abe-421f-add6-971cdd9129ce">
<img width="411" alt="Screenshot 2023-12-01 at 10 36 10 am" src="https://github.com/grafana/oncall/assets/59047654/a55e1c1d-a1aa-4c6d-b05f-12c3a7936711">



## Which issue(s) this PR fixes

Fixes #3407 

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/oncall?pullRequest=3477) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/oncall?pullRequest=3477) before we can accept your contribution.<br/><hr/>**xss** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/oncall?pullRequest=3477) it.</sub>

iskhakov commented:
Thank you for the contribution! Great work! 
We might use step "Notify Team" to notify team's escalation chains. Could you please consider renaming "Notify team" to something more specific, like "Notify all team members" or "Notify all users from team" in the code.

xssfox commented:
Thanks @iskhakov 

I've updated the code to mostly be `notify_team_members` to keep the variables not too long and made the frontend changes a bit more clearer

<img width="766" alt="image" src="https://github.com/grafana/oncall/assets/59047654/f1214737-25ff-496e-b612-26bd7544959a">
<img width="380" alt="image" src="https://github.com/grafana/oncall/assets/59047654/14d75822-9922-48ae-92bb-4973e1eeea3f">

There's currently a test failing `engine/apps/integrations/tests/test_views.py` however it seems to be unrelated to my changes. I performed a `make build`, I suspect my branch doesn't have the required requirement for `DjangoDbBlocker`

xssfox commented:
I think this PR is still valid however I'm not going to put effort into fixing up conflicts if there's no intention to merge

matiasb commented on engine/apps/alerts/tasks/notify_team_members.py:
```diff
@@ -0,0 +1,38 @@
+    reason=None,
+    prevent_posting_to_thread=False,
+    notify_even_acknowledged=False,
+    important=False,
```
I don't see the `important` flag to be set anywhere? I think you need to check the escalation step and set accordingly before queuing the task. Also, maybe re-check what params are used/needed here (I don't think we need them all?)

xssfox replied:
Fixed

matiasb commented on engine/apps/public_api/serializers/escalation_policies.py:
```diff
@@ -280,6 +295,11 @@ def update(self, instance, validated_data):
+                    EscalationPolicy.STEP_NOTIFY_TEAM_MEMBERS,
+                    EscalationPolicy.STEP_NOTIFY_TEAM_MEMBERS_IMPORTANT
+                ]:
+                    instance.notify_to_team_members = None
```
Would you add some test(s) for the changes here?

xssfox replied:
Added in `test_create_escalation_policy_using_notify_team_members`

matiasb commented on engine/apps/alerts/escalation_snapshot/snapshot_classes/escalation_policy_snapshot.py:
```diff
@@ -87,6 +90,7 @@ def __init__(
         self.custom_webhook = custom_webhook
         self.notify_schedule = notify_schedule
         self.notify_to_group = notify_to_group
+        self.notify_to_team_members = notify_to_team_members
```
To be sure, have you checked how this works when deserializing an escalation policy snapshot where the `notify_to_team_members` field does not exist? (we should still be able to deserialize snapshots created before the field was around) Maybe we need a test for this scenario to double-check.

xssfox replied:
I've fixed this along with adding a test for it.

matiasb reviewed (commented):
Thanks (again) for the contribution, I like the idea. Leaving a few comments I think we need to address before merging the changes. Happy to keep reviewing/discussing updates :+1:

xssfox commented:
Thanks for the detailed comments, will have a look in the coming days :)

xssfox commented:
Thanks @matiasb, I've rebased to latest dev, added some tests and I believe fixed the issues above. I've also tested locally to make sure everything still works and it seems to. Let me know if there are any other changes that need looking at.

matiasb commented:
Thanks for the updates, I will be taking a look before the end of the week :+1:

matiasb commented:
Hi @xssfox! I hope you don't mind but it seems I'm not able to push updates to this PR, so forked it [here](https://github.com/grafana/oncall/pull/3908) to implement some additional refactoring and changes. Happy to share any details on the updates, let me know if that looks ok to you. Will get someone else to do a final review, but I think it should be ready to merge with that. Thanks for your contribution, and your patience :+1:

xssfox commented:
@matiasb Looks good, thanks!
